### ISSUE TITLE
Remove regex-based JS/HTML rewriting, switch to AST-based rewriting

### ISSUE THREAD

**Issue by flotwig** [cypress-io/cypress#7297]
AST-based JS/HTML rewriting is available as an experimental feature with https://github.com/cypress-io/cypress/pull/5273

Here are the remaining tasks before it can fully replace the existing regex-based rewriting:

- [ ] Validation with users - is it solving the issues we set out to fix?
- [ ] Rewrite all traffic.
	- may require inlining `resolvers.ts`
- [ ] Do https://github.com/cypress-io/cypress/issues/7268 or have it be fixed by inlining `resolvers.ts`
	- or else rewriting will break if a user visits an http:// url that redirects to https://
- [ ] Inject HTML via `parse5` instead of regex.
- [ ] Resolve issue where proxy will no longer work for normal web browsing outside of Cypress frame. Possible solution: inject a stub `window.Cypress.resolveWindowReference` into every html page here: https://github.com/cypress-io/cypress/blob/2ed56949cb14b4630ca11ec8c4cda715aa9bfa26/packages/server/lib/util/inject.js#L1-L28
- [ ] Performance measurements:
    - [ ] difference between rewriting with sourcemap and without
	- [ ] comparison to regex-based rewriting
- [ ] Performance improvements: 
	- [ ] improve speed: add streaming parsing/rewriting of JS (parse5 handles rewriting HTML as a stream)
	- [ ] mem usage: explore caching js in deferred-sourcemap-cache on filesystem
	- [ ] understand how sourcemaps are cached in browser and how we can reduce unnecessary regeneration/not store old sourcemaps in RAM
- [ ] Add comments around injected code to help users understand - potentially include original code in a comment
- [ ] Need a real fix for Electron worker_threads https://github.com/electron/electron/issues/23366
- [ ] Fix all code TODOs from original PR: https://github.com/cypress-io/cypress/pull/5273

**Comment by drumbeg** [comment]
I am using expermentalSourceRewriting in 4.7.0 and getting the error

```
 /root/.cache/Cypress/4.7.0/Cypress/Cypress[40]: ../../third_party/electron_node/src/node_worker.cc:386:virtual node::worker::Worker::~Worker(): Assertion `thread_joined_' failed.
azure-ci_1   | The Test Runner unexpectedly exited via a exit event with signal SIGABRT
```

Is this a known issue and is there a workaround? This causes our tests to fail in CI even when we have a fully passing suite.

**Comment by flotwig** [comment]
@drumbeg Does it happen while exiting the tests? If so, that is a known bug, caused by some issue in Electron: electron/electron#23366 Cypress uses a workaround, but it seems to be failing in your case. Does it fail every time?

If it happens before the end of the test suite, that's a new bug, in which case please open a new issue and tag me there to track it.

Either way, [debug logs][debug] would be helpful from your test run. You can enable debug logs w/ environment variables:

* On Linux/macOS: `DEBUG=cypress:* cypress ...`
* On Windows: `npx cross-env DEBUG=cypress:* cypress ...`

Make sure to scrub any private data from the logs before sharing here.

[debug]: https://docs.cypress.io/guides/guides/debugging.html#Print-DEBUG-logs

**Comment by drumbeg** [comment]
It only happens when exiting the tests and crucially it only happens when running within a custom built Docker container based on Oracle Linux 7. I get the failure every time, again, only at the end of the suite.

I'm having other issues when trying to attach the debugger.

**Comment by drumbeg** [comment]
@flotwig shall I add debug logs to issue #7572 or are they not useful at this point? Looks like you are underway with looking at a workaround.

**Comment by flotwig** [comment]
@drumbeg yeah, just merged that PR; it may fix the issue for you once released

**Comment by drumbeg** [comment]
4.8.0 looking good 👍
